Refactor Template Resolvers

gsteel opened the pull request:
Amends the `ResolverInterface` to `resolve(non-empty-string): non-empty-string` with failure conditions meaning an exception is thrown.

Resolvers are largely intact but they all now correctly implement the interface and expected behaviour.

TODO:

- [x] Decide how we will expect template maps and directory lists to be configured
- [x] Add factories for the resolvers and wire them up in ConfigProvider
- [x] Decide what fetching `ResolverInterface` from the container will provide by default
- [x] Refactor documentation for v3 with less MVC emphasis

gsteel committed:
Amend the resolver interface to return the template file or throw an exception

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Refactor template map resolver

Adds types and the final keyword.
Refines pairs to non-empty-string

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Refactor `TemplatePathStack` resolver

Adds types and the final keyword.

Refines paths to non-empty-string.

Removes deprecated constants and properties

Removes the methods:

- `setOptions`
- `setDefaultSuffix`
- `getDefaultSuffix`
- `normalizePath` (Is now private)
- `setLfiProtection`
- `isLfiProtectionOn`
- `getLastLookupFailure` (Already deprecated)

Brings behaviour in-line with the interface definition.

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Refactor `PrefixPathStack` resolver

Adds types and the final keyword.

Refines paths to non-empty-string.

Brings behaviour in-line with the interface definition.

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Refactor `AggregateResolver`

Adds types and the final keyword.

Brings behaviour in-line with the interface definition.

Removes `FAILURE_*` constants

Removes deprecated methods

- `getLastSuccessfulResolver`
- `getLastLookupFailure`

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Refactor `RelativeFallbackResolver`

Adds types and the final keyword.

Brings behaviour in-line with the interface definition.

Requires the view model helper as a constructor dependency

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Extend `TemplateCannotBeFound` from `RuntimeException` for fewer BC breaks

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Adjust exception message expectations

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Improve types and docs for the `resolver` method, and invert yoda conditions

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Minor QA

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
The `PrefixPathStackResolver` should accept a default suffix and pass it to the underlying resolver(s)

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Adds factories for the various template resolvers.

Aliases the aggregate resolver to the resolver interface using an MVC-like resolver setup.

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel commented:
This is blocked by #277. The plugin manager must be available in the container in order to test the aggregate resolver factory

gsteel committed:
Adds integration test for the aggregate resolver factory to ensure resolution order is predictable by default.

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Make new factories as internal and import configuration shapes for better type inference

Signed-off-by: George Steel <george@net-glue.co.uk>

Ocramius commented on src/ConfigProvider.php:
```diff
@@ -54,6 +62,45 @@ public function __invoke(): array
+                 * This is conventional for an MVC app
+                 */
+                'template_map' => [
+                    // 'template-name' => 'path/to/template.phtml',
```
Let's use `__DIR__` in the example code

Ocramius commented on src/ConfigProvider.php:
```diff
@@ -54,6 +62,45 @@ public function __invoke(): array
+                 * This is conventional for an MVC app
+                 */
+                'template_path_stack' => [
+                    // 'path/to/a/directory/of/templates',
```
Let's use `__DIR__` in the example code

Ocramius commented on src/ConfigProvider.php:
```diff
@@ -54,6 +62,45 @@ public function __invoke(): array
+                 * Templates configured here will be provided to the PrefixPathStackResolver
+                 */
+                'prefix_template_path_stack' => [
+                    // 'prefix' => 'path/to/a/directory/of/templates',
```
Let's use `__DIR__` in the example code

Ocramius commented on src/ConfigProvider.php:
```diff
@@ -54,6 +62,45 @@ public function __invoke(): array
+                /**
+                 * The default template suffix is configured here in MVC applications
+                 */
+                //'default_template_suffix' => 'phtml',
```
Why is this commented out? We can provide our default :thinking:

gsteel replied:
Because this is potentially configured in `view_manager.default_template_suffix` and `templates.extension`, specifying the default in `templates.extension` will make the former impossible to configure which will be confusing for users.

I thought it better to not _configure_ it at all and rely on a fallback constant for when it's completely undefined

Ocramius commented on src/ConfigProvider.php:
```diff
@@ -54,6 +62,45 @@ public function __invoke(): array
+                 * This is conventional for a Mezzio app
+                 */
+                'map' => [
+                    // 'template-name' => 'path/to/template.phtml',
```
Let's use `__DIR__` in the example code

Ocramius commented on src/ConfigProvider.php:
```diff
@@ -54,6 +62,45 @@ public function __invoke(): array
+                /**
+                 * The default template filename extension is configured here. The default is 'phtml'
+                 */
+                //'extension' => 'phtml',
```
Why is this commented out? We can provide our default 🤔

Ocramius commented on src/Helper/Service/Configuration.php:
```diff
@@ -51,4 +53,22 @@ public static function viewEncoding(
+     */
+    public static function defaultTemplateSuffix(
+        ContainerInterface $container,
+        string $fallbackDefault = self::DEFAULT_TEMPLATE_SUFFIX,
```
this parameter is never set by anyone: let's kill it

Ocramius commented on src/Resolver/AggregateResolver.php:
```diff
@@ -72,77 +39,27 @@ public function count()
-                $this->lastSuccessfulResolver = $resolver;
-                return $resource;
+            try {
+                return $resolver->resolve($name);
+            } catch (TemplateCannotBeFound) {
+                continue;
```
Note: using exceptions for this can be really expensive, and this is critical path for most apps.

gsteel replied:
👍 I've added, used and tested `ResolverInterface::has($name): bool`

Ocramius reviewed (changes requested):
Only skimmed this: I think the exception used for flow control is something we should avoid here.

A `TemplatePathStack` resolver (used by default by most apps) will really suffer here, IMO.

gsteel committed:
Add a `has(string): bool` method to the template `ResolverInterface`

Using `has` in place of `try/catch` will be more performant for an especially hot path in most apps

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Use the `__DIR__` magic constant in variable examples

Signed-off-by: George Steel <george@net-glue.co.uk>

gsteel committed:
Remove unused parameter

Signed-off-by: George Steel <george@net-glue.co.uk>

Ocramius commented on src/Resolver/AggregateResolver.php:
```diff
@@ -53,13 +53,24 @@ public function attach(ResolverInterface $resolver, int $priority = 1): self
-            try {
-                return $resolver->resolve($name);
-            } catch (TemplateCannotBeFound) {
+            if (! $resolver->has($name)) {
```
This is now worse, because we do double the I/O 🤣

Sorry!

gsteel replied:
I/O is only happening in the PathStack variant.

What do you think is a good solution here?

Internal in-memory cache? Or changing `Resolver::get(string): string|false`?

Ocramius replied:
> I/O is only happening in the PathStack variant.

Which is the most common one, heh

I think sticking to `|false` is the least evil change right now

gsteel committed:
Revert addition of `ResolverInterface::has()`. Revert behaviour of `ResolverInterface::resolve()` to return `string|false` instead of throwing an exception.

Signed-off-by: George Steel <george@net-glue.co.uk>

Ocramius reviewed (approved):
Only managed to skim the patch, as I'm on the run - a second review by someone else may be a good idea

froschdesign commented on src/Resolver/TemplateMapResolver.php:
```diff
@@ -174,21 +147,15 @@ public function get($name)
+    public function resolve(string $name): string|false
     {
-        return $this->get($name);
+        return $this->map[$name] ?? false;
```
The method `get` uses `has` and throws an exception, but is `false` the desired result here?

gsteel replied:
Yes, all implementations of `resolve` have the signature `resolve(non-empty-string): non-empty-string|false`.

I initially returned only string or throw, but Marco pointed out the performance impact of throwing on a hot path so reverted to `string|false`

Both `has` and `get` are existing methods on this class - I would have removed them but couldn't find a convincing reason to break even more BC apart from immutability for the sake of it.

froschdesign replied:
> but Marco pointed out the performance impact of throwing on a hot path

Good point! 👍🏻
